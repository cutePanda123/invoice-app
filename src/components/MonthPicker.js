import React from 'react';
import PropTypes from 'prop-types';
import Utility from '../utilility';

class MonthPicker extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            isOpen: false,
            selectedYear: props.year,
            selectedMonth: props.month
        };
    }

    onYearChange = (event, year) => {
        event.preventDefault();
        this.setState({
            selectedYear: year
        });
    }

    onMonthChange = (event, month) => {
        event.preventDefault();
        this.setState({
            selectedMonth: month,
            isOpen: false
        });
        this.props.onDateChange(this.state.selectedYear, this.state.selectedMonth);
    }


    onDropdownClick = (event) => {
        event.preventDefault();
        this.setState({
            isOpen: !this.state.isOpen
        });
    }

    render() {
        const {selectedMonth, selectedYear} = this.state;
        const {year} = this.props;
        const { isOpen } = this.state;
        const monthRange = Utility.range(12, 1);
        const yearRange = Utility.range(9, -4).map(number => number + year);
        return (
            <div className="dropdown month-picker-component">
                <h4>Select a month</h4>
                <button
                    className="btn btn-lg btn-secondary dropdown-toggle"
                    onClick={this.onDropdownClick}
                >
                    {`${Utility.getMonthString(selectedMonth)}/${selectedYear}`}
                </button>
                {
                    isOpen &&
                    <div className="dropdown-menu" style={{display: 'block'}}>
                        <div className="row">
                            <div className="col border-right">
                                <a key="0" className="dropdown-item">Year</a>
                                {yearRange.map((curYear, idx) => {
                                    return (
                                        <a
                                            href="#"
                                            key={idx + 1} 
                                            className={curYear === selectedYear ? "dropdown-item active" : "dropdown-item"}
                                            onClick={
                                                (event) => {
                                                    this.onYearChange(event, curYear);
                                                }
                                            }
                                        >
                                            {curYear}
                                        </a>
                                    );
                                })}
                            </div>
                            <div className="col">
                                <a key="0" className="dropdown-item">Month</a>
                                {monthRange.map((curMonth, idx) => {
                                    return (
                                        <a
                                            href="#"
                                            key={idx + 1} 
                                            className={curMonth === selectedMonth ? "dropdown-item active" : "dropdown-item"}
                                            onClick={
                                                (event) => {
                                                    this.onMonthChange(event, curMonth);
                                                }
                                            }
                                        >
                                            {Utility.getMonthString(curMonth)}
                                        </a>
                                    );
                                })}
                            </div>
                        </div>
                    </div>
                }
            </div>
        );
    }
}

MonthPicker.propTypes = {
    year: PropTypes.number.isRequired,
    month: PropTypes.number.isRequired,
    onDateChange: PropTypes.func.isRequired
}

MonthPicker.defaultProps = {
    year: 1970,
    month: 1
}

export default MonthPicker;