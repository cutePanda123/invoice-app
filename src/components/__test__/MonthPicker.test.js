import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import MonthPicker from '../MonthPicker';

let props = {
    year: 2020,
    month: 8,
    onDateChange: jest.fn()
};

let wrapper = null;

describe('MonthPicker test', () => {
    beforeEach(() => {
        wrapper = mount(<MonthPicker {...props} />);
    });

    it('snapshot test', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should render correct year and month', () => {
        expect(wrapper.find('.dropdown-toggle').first().text()).toEqual('08/2020');
        expect(wrapper.find('dropdown-menu').length).toEqual(0);
        expect(wrapper.state('isOpen')).toEqual(false);
    });

    it('should show dropdown list on dropdown toggle click', () => {
        wrapper.find('.dropdown-toggle').simulate('click');
        expect(wrapper.state('isOpen')).toEqual(true);
        expect(wrapper.find('.dropdown-menu').length).toEqual(1);
        expect(wrapper.find('.years-range .dropdown-item').length).toEqual(9);
        expect(wrapper.find('.months-range .dropdown-item').length).toEqual(12);
        expect(wrapper.find('.years-range .active').text() * 1).toEqual(2020);
        expect(wrapper.find('.months-range .active').text()).toEqual("08");
        expect(wrapper.find('.years-range .dropdown-item').first().text()).toEqual("2016");
        expect(wrapper.find('.months-range .dropdown-item').first().text()).toEqual("01");
    });

    it('click year should trigger correct handler', () => {
        wrapper.find('.dropdown-toggle').simulate('click');
        wrapper.find('.years-range .dropdown-item')
            .first()
            .simulate('click', { 
                preventDefault: () => {},
                stopPropagation: () => {}
            });
        expect(props.onDateChange).toHaveBeenCalledWith(2016, 8);
    });

    it('click month should trigger correct handler', () => {
        wrapper.find('.dropdown-toggle').simulate('click');
        wrapper.find('.months-range .dropdown-item')
            .first()
            .simulate('click', { 
                preventDefault: () => {},
                stopPropagation: () => {}
            });
        expect(props.onDateChange).toHaveBeenCalledWith(2020, 1);
        expect(wrapper.state('isOpen')).toEqual(false);
    });

    it('click dropdown itself cannot close the dropdown if it is open', () => {
        let clickEventCallback = null;
        document.addEventListener = jest.fn((event, cb) => {
            clickEventCallback = cb;
        });
        let localWrapper = mount(<MonthPicker {...props} />);
        localWrapper.find('.dropdown-toggle').simulate('click');
        clickEventCallback({
            target: ReactDOM.findDOMNode(localWrapper.instance())
        });
        expect(localWrapper.state('isOpen')).toEqual(true);
    });

    it('click document can close the dropdown if it is open', () => {
        let clickEventCallback = null;
        document.addEventListener = jest.fn((event, cb) => {
            clickEventCallback = cb;
        });
        let localWrapper = mount(<MonthPicker {...props} />);
        localWrapper.find('.dropdown-toggle').simulate('click');
        clickEventCallback({
            target: document
        });
        expect(localWrapper.state('isOpen')).toEqual(false);
    });
});