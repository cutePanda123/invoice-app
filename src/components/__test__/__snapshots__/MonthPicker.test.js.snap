// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthPicker test snapshot test 1`] = `
<MonthPicker
  month={8}
  onDateChange={[MockFunction]}
  year={2020}
>
  <div
    className="dropdown month-picker-component"
  >
    <h4>
      Select a month
    </h4>
    <button
      className="btn btn-lg btn-secondary dropdown-toggle"
      onClick={[Function]}
    >
      08/2020
    </button>
  </div>
</MonthPicker>
`;
